### PR TITLE
Support embed binary file into custom section

### PR DESCRIPTION
**Pull request by zhuangton**
Support embed binary file into custom section:
Add (-p + binaryfile name) (-c + cutsom section name) conditional options to insert binary file data into custom fields
Testing method:
Execute the wasmscript.sh script,Complete inserting bin file data into wasm file

### CODE REVIEW

File: src/tools/wat2wasm.cc [hunk 1]
```diff
@@ -139,7 +150,6 @@ int ProgramMain(int argc, char** argv) {
   if (Failed(result)) {
     WABT_FATAL("unable to read file: %s\n", s_infile);
   }
-
```
Comment (sbc100) [inline src/tools/wat2wasm.cc:142]: Unrelated whitespace change (revert?)

File: src/tools/wat2wasm.cc [hunk 2]
```diff
@@ -150,6 +160,17 @@ int ProgramMain(int argc, char** argv) {
...
   if (Succeeded(result)) {
+    if (s_read_binfile) {
+        std::ifstream file(s_binfile, std::ios::binary);
+        if (file) {
+          std::vector<uint8_t> custom_data(std::istreambuf_iterator<char>(file), {});
+          module->customs.emplace_back(Location(0), s_custom_name, custom_data);
+        }
+        else {
+          WABT_FATAL("unable to read binary file\n");
+        }
```
Comment (sbc100) [inline src/tools/wat2wasm.cc:171]: The above block looks like its indented by 4 instead of 2 spaces

File: src/tools/wat2wasm.cc [hunk 3]
```diff
@@ -150,6 +160,17 @@ int ProgramMain(int argc, char** argv) {
   }
 
   if (Succeeded(result)) {
+    if (s_read_binfile) {
+        std::ifstream file(s_binfile, std::ios::binary);
+        if (file) {
+          std::vector<uint8_t> custom_data(std::istreambuf_iterator<char>(file), {});
+          module->customs.emplace_back(Location(0), s_custom_name, custom_data);
+        }
+        else {
```
Comment (sbc100) [inline src/tools/wat2wasm.cc:169]: `else` goes on the line along with the closing brace above

File: test/custom-write-binary/wasmscript.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
Comment (sbc100) [inline test/custom-write-binary/wasmscript.sh:1]: Can we give this shell script a better name?

Also, we probably can't have shell scripts in the test setup since we need to be able to run the full suite under windows.

File: src/tools/wat2wasm.cc [hunk 5]
```diff
@@ -80,6 +85,12 @@ static void ParseOptions(int argc, char* argv[]) {
   parser.AddOption('o', "output", "FILE",
                    "Output wasm binary file. Use \"-\" to write to stdout.",
                    [](const char* argument) { s_outfile = argument; });
+  parser.AddOption('p', "inputfile", "PATHFILE",
+                   "Input binary file. Use \"-\" to get data.",
```
Comment (sbc100) [inline src/tools/wat2wasm.cc:89]: How about "Input binary for custom section" for the docs and "custom_section_filename" for the name?

File: src/tools/wat2wasm.cc [hunk 6]
```diff
@@ -80,6 +85,12 @@ static void ParseOptions(int argc, char* argv[]) {
...
+                   [](const char* argument) { s_binfile = argument; s_read_binfile = true; });
+  parser.AddOption('c', "customname", "CUSTOM",
+                   "Input custom section name. Use \"-\" to write custom section.",
```
Comment (sbc100) [inline src/tools/wat2wasm.cc:92]: How about calling this `--append-section`?      (Since it can only append the end).

What does `Use \"-\" to write custom section` mean here?

**Review by sbc100 — COMMENTED** [review]
Can you explain a little more about your use case here?

I guess the text format cannot yet express custom sections (See https://github.com/WebAssembly/design/issues/1153).  The final comments on that issue suggest that maybe the annotations proposal can to this?

**Comment by keithw** [comment]
This seems like useful functionality, but could we do it in a separate command-line program that lets the user add/print/remove a custom section, instead of adding more flags to wat2wasm/wasm2wat?

(I think we want to keep those tools fairly focused on assembly/disassembly if the plan is to migrate over to the wasm-tools or binaryen codebases -- the ability to edit a module or extract pieces seems like a different thing.)

**Comment by sbc100** [comment]
I tend to agree.  FWIW I think llvm-objcopy can already do this?  Is that right @dschuff ?

**Comment by dschuff** [comment]
Yes, llvm-objcopy can add or remove custom sections, to and from files (I'm not sure it can e.g. print them to stdout though).

**Comment by zhuangton** [comment]
Thank you for your guidance. I have completed using llvm-objcopy
